### ISSUE TITLE
Error: Cannot read property 'type' of undefined

### ISSUE THREAD

**Issue by seeden** [yannickcr/eslint-plugin-react#182]
When I will run 
> node ./node_modules/eslint/bin/eslint.js --ext .js,.jsx ./src/

I can see this error:
```sh
D:\git\react-form-controlled\node_modules\eslint-plugin-react\lib\rules\prop-typ
es.js:469
          propTypes.parent.type !== 'AssignmentExpression' &&
                          ^
TypeError: Cannot read property 'type' of undefined
    at markPropTypesAsDeclared (D:\git\react-form-controlled\node_modules\eslint
-plugin-react\lib\rules\prop-types.js:469:27)
    at EventEmitter.ClassProperty (D:\git\react-form-controlled\node_modules\esl
int-plugin-react\lib\rules\prop-types.js:536:7)
    at EventEmitter.emit (events.js:107:17)
    at Controller.controller.traverse.enter (D:\git\react-form-controlled\node_m
odules\eslint\lib\eslint.js:794:25)
    at Controller.__execute (D:\git\react-form-controlled\node_modules\eslint\no
de_modules\estraverse\estraverse.js:397:31)
    at Controller.traverse (D:\git\react-form-controlled\node_modules\eslint\nod
e_modules\estraverse\estraverse.js:495:28)
    at EventEmitter.module.exports.api.verify (D:\git\react-form-controlled\node
_modules\eslint\lib\eslint.js:787:24)
    at processText (D:\git\react-form-controlled\node_modules\eslint\lib\cli-eng
ine.js:199:27)
    at processFile (D:\git\react-form-controlled\node_modules\eslint\lib\cli-eng
ine.js:224:12)
    at D:\git\react-form-controlled\node_modules\eslint\lib\cli-engine.js:320:26
```

parent is not defined. When I changed the condition to:
```js
propTypes && propTypes.parent &&
          propTypes.parent.type !== 'AssignmentExpression' &&
          propTypes.property &&
          curDeclaredPropTypes
```

I can see same error few lines later:
```sh
> node ./node_modules/eslint/bin/eslint.js --ext .js,.jsx ./src/
```

```sh
D:\git\react-form-controlled\node_modules\eslint-plugin-react\lib\rules\prop-typ
es.js:487
            buildReactDeclarationTypes(propTypes.parent.right);
                                                       ^
TypeError: Cannot read property 'right' of undefined
    at markPropTypesAsDeclared (D:\git\react-form-controlled\node_modules\eslint
-plugin-react\lib\rules\prop-types.js:487:56)
    at EventEmitter.ClassProperty (D:\git\react-form-controlled\node_modules\esl
int-plugin-react\lib\rules\prop-types.js:538:7)
    at EventEmitter.emit (events.js:107:17)
    at Controller.controller.traverse.enter (D:\git\react-form-controlled\node_m
odules\eslint\lib\eslint.js:794:25)
    at Controller.__execute (D:\git\react-form-controlled\node_modules\eslint\no
de_modules\estraverse\estraverse.js:397:31)
    at Controller.traverse (D:\git\react-form-controlled\node_modules\eslint\nod
e_modules\estraverse\estraverse.js:495:28)
    at EventEmitter.module.exports.api.verify (D:\git\react-form-controlled\node
_modules\eslint\lib\eslint.js:787:24)
    at processText (D:\git\react-form-controlled\node_modules\eslint\lib\cli-eng
ine.js:199:27)
    at processFile (D:\git\react-form-controlled\node_modules\eslint\lib\cli-eng
ine.js:224:12)
    at D:\git\react-form-controlled\node_modules\eslint\lib\cli-engine.js:320:26
```

After change condition again to
```js
if (propTypes && propTypes.property && propTypes.parent) {
```

it is working all right

**Comment by vilppuvuorinen** [comment]
I run into the same issue while working with `feross/standard`. I ended up adding a `propTypes.parent` check to the three condition blocks two of which are described above.

**Comment by yannickcr** [comment]
Can you give me the code that crash the rule ?

**Comment by vilppuvuorinen** [comment]
Unfortunately at least I am unable to provide that. I cannot find a commit where the error was present and the lacking verbosity would have made it almost impossible to narrow the responsible code down. However, I had to disable linting for a couple of weeks due to this error. Luckily it is quite simple to fix by fiddling through node_modules and does not seem to be present currently.

**Comment by yannickcr** [comment]
Ok, so I will fix it with a blind bugfix, not very clean but at least it should solve your problem.

**Comment by janpaul123** [comment]
I have another case that triggers this error message:

```js
const SomeComponent = React.createClass({
  propTypes: SomeOtherComponent.propTypes,
});
```

This is a common pattern in our codebase for wrapping components that transparently pass through properties to a child component.

**Comment by yannickcr** [comment]
Thanks @janpaul123, I added this pattern to the test suite.

This should be fixed in `v3.2.1`

**Comment by janpaul123** [comment]
Sweet, thanks!
